### PR TITLE
Implements the '/users' API endpoint for saving and retrieving user's details  using the Users Authentication API.

### PR DESCRIPTION
**Pull request by Cayle**

### CODE REVIEW

File: capstone/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 1]
```diff
@@ -1,11 +1,218 @@
...
+      String user_status = Boolean.toString(isUserLoggedIn);
+      userDetails.put("user_status", user_status);
+      String loginUrl = userService.createLoginURL(
+        urlToRedirectToAfterUserLogsIn
+      );
+      userDetails.put("loginUrl", loginUrl);
+      userDetails.put("status_code", "401");
+      return userDetails;
+    }
+  }
```
Comment (rayliu0317) [inline capstone/src/main/java/com/google/sps/servlets/UserServlet.java:129]: This whole block is kind of hard to follow mainly due to that there are too many nested blocks (if -> if -> try -> try -> else -> do/while).

You can try to refactor this by putting each separate logic into helper functions and return early when an if condition is not satisfied. Example:
if (!condition) {
 // work
 return
}
// work

will look much nicer than
if (condition) {
  work
} else {
  work
}

Comment (Cayle) [reply capstone/src/main/java/com/google/sps/servlets/UserServlet.java]: Done, you can review it.

File: capstone/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 2]
```diff
@@ -1,11 +1,218 @@
...
 import javax.servlet.http.HttpServletResponse;
+import javax.sql.DataSource;
 
 @WebServlet("/users")
 public class UserServlet extends HttpServlet {
-    //TODO doGet for getting user info
-    //TODO doPost for creating and storing a user account
-}
\ No newline at end of file
+  private final String get_user_row_query =
```
[the remark this answers is not in the document]
Comment (Cayle) [reply capstone/src/main/java/com/google/sps/servlets/UserServlet.java]: Done

File: capstone/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 3]
```diff
@@ -1,11 +1,269 @@
...
+            String user_name = result.getString("user_name");
+            userDetails.put("user_name", user_name);
+            String university = result.getString("name");
+            userDetails.put("university", university);
+            String logoutUrl = userService.createLogoutURL(
+              URL_TO_REDIRECT_AFTER_USER_LOGS_OUT
+            );
+            userDetails.put("logoutUrl", logoutUrl);
+            userDetails.put("status_code", "200");
+          } while (result.next());
```
Comment (rayliu0317) [inline capstone/src/main/java/com/google/sps/servlets/UserServlet.java:173]: This should be pulled out to a separate function also.

The current layout is still too nested.

File: capstone/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 4]
```diff
@@ -1,11 +1,269 @@
...
+      ) {
+        queryStatement.setString(1, email);
+        ResultSet result = queryStatement.executeQuery();
+        if (result.next() == false) {
+          response.sendRedirect("/userDetails.html");
+          userDetails.put("status_code", "307");
+          return userDetails;
+        } else {
+          do {
+            String user_status = Boolean.toString(isUserLoggedIn);
```
Comment (rayliu0317) [inline capstone/src/main/java/com/google/sps/servlets/UserServlet.java:157]: inline this to the line below will make this shorter.

Comment (rayliu0317) [reply capstone/src/main/java/com/google/sps/servlets/UserServlet.java:157]: And same for all below.

File: capstone/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 5]
```diff
@@ -1,11 +1,269 @@
...
+
+    UserService userService = UserServiceFactory.getUserService();
+    try {
+      HashMap<String, String> userDetails = getUserDetails(
+        pool,
+        userService,
+        GET_USER_ROW_QUERY,
+        response
+      );
+      String userResult = new Gson().toJson(userDetails);
```
Comment (rayliu0317) [inline capstone/src/main/java/com/google/sps/servlets/UserServlet.java:201]: move these out of try.

File: capstone/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 6]
```diff
@@ -1,11 +1,269 @@
...
+      response.getWriter().println(userResult);
+    } catch (SQLException ex) {
+      throw new RuntimeException(
+        "There is an error with your sql statement ... ",
+        ex
+      );
```
Comment (rayliu0317) [inline capstone/src/main/java/com/google/sps/servlets/UserServlet.java:207]: these can be in one line I think.

File: capstone/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 7]
```diff
@@ -1,11 +1,269 @@
...
+  )
+    throws SQLException {
+    try (Connection conn = pool.getConnection()) {
+      try (
+        PreparedStatement queryStatement = conn.prepareStatement(
+          create_new_user_query
+        )
+      ) {
+        String full_name = request.getParameter("full_name");
+        String user_name = request.getParameter("user_name");
```
Comment (rayliu0317) [inline capstone/src/main/java/com/google/sps/servlets/UserServlet.java:233]: Inline these where you use them.

File: capstone/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 8]
```diff
@@ -1,11 +1,269 @@
...
+  )
+    throws SQLException, IOException {
+    HashMap<String, String> userDetails = new HashMap<>();
+    String email = userService.getCurrentUser().getEmail();
+    Boolean isUserLoggedIn = userService.isUserLoggedIn();
+    try (Connection conn = pool.getConnection()) {
+      try (
+        PreparedStatement queryStatement = conn.prepareStatement(
+          get_user_row_query
+        )
```
Comment (rayliu0317) [inline capstone/src/main/java/com/google/sps/servlets/UserServlet.java:147]: combine this with the try above:
try(Connection conn = pool.getConnection();
      PreparedStatement queryStatement = conn.prepareStatement(
          get_user_row_query) {
}

File: capstone/src/main/java/com/google/sps/servlets/UserServlet.java (hunk 4 above)
Comment (Cayle) [reply capstone/src/main/java/com/google/sps/servlets/UserServlet.java:157]: By inline, do you mean that rather than using two lines, I should do something like userDetails.put("user_status", Boolean.toString(isUserLoggedIn)); ?

Comment (rayliu0317) [reply capstone/src/main/java/com/google/sps/servlets/UserServlet.java:157]: yes. no need to create a new variable.

File: capstone/src/main/java/com/google/sps/servlets/UserServlet.java (hunk 5 above)
Comment (Cayle) [reply capstone/src/main/java/com/google/sps/servlets/UserServlet.java:201]: I don't really get this, can you explain this more?

Comment (Cayle) [reply capstone/src/main/java/com/google/sps/servlets/UserServlet.java:201]: I don't know but, if I move line 201 out of the try block, I won't be able to access the userDetails created.
